### PR TITLE
Support inserting and replaying router expert decisions in trainer

### PR DESCRIPTION
**Pull request by khatwanimohit**
# Description

- In train.py loss_fn, extract forced_routed_experts from batch data and pass to model forward
- In RoutedMoE, override top_k_indices and route tokens with forced_routed_experts across sparse_matmul and dense_matmul
- In Decoder and NNXDecoder, plumb forced_routed_experts down across scanned/sequential decoder layers
- Plumbed forced_routed_experts across model implementations (Qwen3, Qwen3.5, DeepSeek, Gemma4, Mixtral)
- Added unit tests in tests/unit/forced_routing_test.py
- Added integration test in tests/test_trainer_router_replay.py verifying loss computation with forced routing

You can also provide a comma-separated list. If you don't want to close a bug but
simply to reference it, use BUGS, e.g.: 
BUGS: b/536114593

*Notice 1:* Once all tests pass, the "pull ready" label will automatically be assigned.
This label is used for administrative purposes. Please do not add it manually.

*Notice 2:* For external contributions, our settings currently require an approval from a MaxText maintainer to trigger CI tests.

# Tests

unit tests

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/layers/moe.py [hunk 1]
```diff
@@ -977,15 +1000,23 @@ def permute(
 
       if roll_to_expert_id is not None:
         flatten_selected_experts = (flatten_selected_experts - roll_to_expert_id) % self.num_experts
-      sorted_selected_experts = jnp.argsort(flatten_selected_experts)
+      if forced_routed_experts is not None:
+        # Fix padding bug: map -1 to dummy valid indices to distribute load
+        valid_mask = flatten_selected_experts >= 0
+        dummy_indices = jnp.arange(flatten_selected_experts.shape[0]) % self.num_experts
+        flatten_selected_experts_safe = jnp.where(valid_mask, flatten_selected_experts, dummy_indices)
```
[the remark this answers is not in the document]
Comment (khatwanimohit) [reply src/maxtext/layers/moe.py:1033]: done

File: src/maxtext/training_engine/maxtext_engine.py [hunk 2]
```diff
@@ -145,6 +146,108 @@ def _batch_signature(dynamic_batch: Any, static_batch: dict[str, Any]) -> Any:
...
+      "inputs": token_ids,
+      "inputs_position": positions,
+      "inputs_segmentation": segment_ids,
+      "targets": jnp.roll(token_ids, -1, axis=-1),
+      "targets_segmentation": token_mask,
+  }
+  forced_routed_experts = getattr(payload, "forced_routed_experts", None)
+  if forced_routed_experts is not None:
+    kwargs["forced_routed_experts"] = forced_routed_experts
+  return kwargs
```
[the remark this answers is not in the document]
Comment (khatwanimohit) [reply src/maxtext/training_engine/maxtext_engine.py:220]: done

File: src/maxtext/layers/moe.py [hunk 3]
```diff
@@ -738,47 +738,68 @@ def should_update_load_balance(self):
...
 
     if self.config.decoder_block in (ctypes.DecoderBlockType.DEEPSEEK, ctypes.DecoderBlockType.DEEPSEEK4):
       top_k_weights = self.deepseek_scale_weights(top_k_weights)
     else:
       if self.config.decoder_block not in (ctypes.DecoderBlockType.LLAMA4, ctypes.DecoderBlockType.GEMMA4):
         top_k_weights = jax.nn.softmax(top_k_weights.astype(jnp.float32), axis=-1).astype(self.dtype)
 
-      # Normalization of router weights (e.g. used by Qwen3, Gemma4).
-      if self.config.norm_topk_prob:
-        top_k_weights /= top_k_weights.sum(axis=-1, keepdims=True)
```
Comment (NuojCheng) [inline src/maxtext/layers/moe.py:781]: The indentation of 779-781 unmatches with line 795-802, please double-check

Comment (NuojCheng) [reply src/maxtext/layers/moe.py:781]: they should be in the else path, right?

File: src/maxtext/layers/decoders.py [hunk 4]
```diff
@@ -1246,17 +1266,47 @@ def __call__(
...
+                DecoderBlockType.QWEN3_5,
+                DecoderBlockType.QWEN3_CUSTOM_MOE,
+                DecoderBlockType.GEMMA4,
+            ):
+              is_moe = True
+            elif cfg.decoder_block in (DecoderBlockType.LLAMA4, DecoderBlockType.ENVY):
+              is_moe = llama4.determine_is_moe_layer(lyr, self.config.interleave_moe_layer_step)
+
+            if is_moe and forced_routed_experts is not None:
+              if forced_routed_experts.ndim == 4:
```
Comment (entrpn) [inline src/maxtext/layers/decoders.py]: Add assertion logic to check that ndim is the value we expect (i.e, 3 or 4), in the case it is sent the wrong dim (1, 2, or 5)? 

same in nnx_decoders if the logic is there, I couldn't find it.

File: src/maxtext/layers/moe.py (hunk 3 above)
Comment (khatwanimohit) [reply src/maxtext/layers/moe.py:781]: fixed

File: src/maxtext/layers/decoders.py (hunk 4 above)
Comment (khatwanimohit) [reply src/maxtext/layers/decoders.py]: added tests

**Review by NuojCheng — APPROVED** [review]

**Review by entrpn — APPROVED** [review]
